codec/src/types/primitives.rs

```diff
@@ -145,19 +186,29 @@ impl_codec_for_tuple!(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11);
 // Vec implementation
 impl<T: Codec> Codec for Vec<T> {
     #[inline]
-    fn write(&self, writer: &mut impl Writer) {
-        writer.write_vec(self);
+    fn write(&self, buf: &mut impl BufMut) {
+        let len = u32::try_from(self.len()).expect("Vec length exceeds u32");
+        varint::write(len, buf);
+        for item in self {
+            item.write(buf);
+        }
     }
 
     #[inline]
     fn len_encoded(&self) -> usize {
-        let len = varint::varint_size(self.len() as u64);
-        self.iter().map(Codec::len_encoded).sum::<usize>() + len
+        let len = u32::try_from(self.len()).expect("Vec length exceeds u32");
+        varint::size(len) + self.iter().map(Codec::len_encoded).sum::<usize>()
     }
 
     #[inline]
-    fn read(reader: &mut impl Reader) -> Result<Self, Error> {
-        reader.read_vec()
+    fn read(buf: &mut impl Buf) -> Result<Self, Error> {
+        let len32 = varint::read::<u32>(buf)?;
+        let len = usize::try_from(len32).unwrap();
```

Comment:
Might as well return an error here imo with `.map_err()` (should not happen but 🤷 )